examples/agents/trainer/train.py

```diff
@@ -0,0 +1,347 @@
+
+  run_config = tf.contrib.learn.RunConfig()
+
+  server_def = tf.train.ServerDef(
+      cluster=run_config.cluster_spec.as_cluster_def(),
+      protocol="grpc",
+      job_name=run_config.task_type,
+      task_index=run_config.task_id)
+
+  server = tf.train.Server(server_def)
+
+  tf.reset_default_graph()
+
+  if agents_config.update_every % agents_config.num_agents:
+    tf.logging.warn('Number of agents should divide episodes per update.')
+
+  worker_device = "/job:%s/task:%d" % (run_config.task_type,
+                                       run_config.task_id)
+
+  tf.logging.debug('is_chief: %s' % run_config.is_chief)
+  tf.logging.debug(pprint.pprint(run_config.cluster_spec.as_dict()))
+  tf.logging.debug('server.target: %s' % server.target)
+  tf.logging.debug('worker_device: %s' % worker_device)
+
+  # By default, all parameters are shared.
+  with tf.device(
+      tf.train.replica_device_setter(
+          worker_device=worker_device,
+          ps_device="/job:ps",
+          cluster=run_config.cluster_spec)):
+
+    batch_env = utility.define_batch_env(
+        lambda: _create_environment(agents_config),
+        agents_config.num_agents, env_processes)
+
+    graph = define_simulation_graph(
+        batch_env, agents_config.algorithm, agents_config)
+
+    loop = _define_loop(
+        graph, log_dir,
+        agents_config.update_every * agents_config.max_length,
+        agents_config.eval_episodes * agents_config.max_length)
+
+    total_steps = int(
+        agents_config.steps / agents_config.update_every *
+        (agents_config.update_every + agents_config.eval_episodes))
+
+    # Exclude episode related variables since the Python state of environments is
+    # not checkpointed and thus new episodes start after resuming.
+    # TODO: Is this preventing these variables from being initialized on the
+    # workers?
+    # saver = utility.define_saver()
+    saver = utility.define_saver(exclude=(r'.*_temporary/.*',))
+
+    device_filters = ["/job:ps",
+                      "/job:%s/task:%d" % (run_config.task_type, run_config.task_id)]
+    # tf.logging.debug('device_filters: %s' % device_filters)
+
+    sess_config = tf.ConfigProto(allow_soft_placement=True,
+                                 device_filters=device_filters
+                                 )
+    if FLAGS.log_device_placement:
+      sess_config.log_device_placement = True
+
+    sess_config.gpu_options.allow_growth = True
+
+    opt = graph.algo._optimizer
+
+    # init_op = tf.group(
+    #     tf.local_variables_initializer(),
+    #     tf.global_variables_initializer())
+    init_op = tf.global_variables_initializer()
+
+    local_init_op = tf.local_variables_initializer()
+
+    if FLAGS.sync_replicas:
+      # Get a reference to the propper local_init_op
+      ready_for_local_init_op = opt.ready_for_local_init_op
+      local_init_op = opt.local_step_init_op
+      if run_config.is_chief:
+        local_init_op = opt.chief_init_op
```

Comment:
Why doesn't the chief initialize local ops?

In https://github.com/tensorflow/agents/issues/17#issuecomment-350505347 you posted the following error
>
```
 INFO:tensorflow:Waiting for model to be ready.  Ready_for_local_init_op:
  Variables not initialized: ppo_temporary/episodes/Variable,
  ppo_temporary/episodes/Variable_1, ppo_temporary/episodes/Variable_2,
  ppo_temporary/episodes/Variable_3, ppo_temporary/episodes/Variable_4,
  ppo_temporary/episodes/Variable_5, ppo_temporary/last_action,
  ppo_temporary/last_mean, ppo_temporary/last_logstd, ready: None
```
That error message is coming from here.
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/session_manager.py#L417

Is the error message happening on the master? I'd imagine that if the master is acting as a worker then it also has local variables that need to be initialized.

Reply:
The error message isn't occurring on the master which is proceeding with training normally except for seemingly repeatedly starting the master session?

```bash
--------------------------------------------------
Phase train (phase step 0, global step 0).
2017-12-13 16:33:46.571102: I tensorflow/core/distributed_runtime/master_session.cc:1004] Start master session df5b8121a99e0179 with config: gpu_options { allow_growth: true } allow_soft_placement: true
2017-12-13 16:33:46.768473: I tensorflow/core/distributed_runtime/master_session.cc:1004] Start master session a7f4259ee7b2e152 with config: gpu_options { allow_growth: true } allow_soft_placement: true
2017-12-13 16:33:46.957017: I tensorflow/core/distributed_runtime/master_session.cc:1004] Start master session 0d3890a298f3ddb3 with config: gpu_options { allow_growth: true } allow_soft_placement: true
INFO:tensorflow:global_step/sec: 13.5033
INFO:tensorflow:global_step/sec: 27.8826
INFO:tensorflow:global_step/sec: 28.0781
2017-12-13 16:34:17.244879: I tensorflow/core/distributed_runtime/master_session.cc:1004] Start master session 041e61fded948009 with config: gpu_options { allow_growth: true } allow_soft_placement: true
2017-12-13 16:34:17.559273: I tensorflow/core/distributed_runtime/master_session.cc:1004] Start master session 73a4b1f71bf752f7 with config: gpu_options { allow_growth: true } allow_soft_placement: true
2017-12-13 16:34:17.678367: I tensorflow/core/distributed_runtime/master_session.cc:1004] Start master session 745a9dc24252072b with config: gpu_options { allow_growth: true } allow_soft_placement: true
INFO:tensorflow:global_step/sec: 27.0679
INFO:tensorflow:global_step/sec: 26.45
INFO:tensorflow:global_step/sec: 25.8405
2017-12-13 16:34:47.918164: I tensorflow/core/distributed_runtime/master_session.cc:1004] Start master session bb207b634fca4147 with config: gpu_options { allow_growth: true } allow_soft_placement: true
2017-12-13 16:34:48.307535: I tensorflow/core/distributed_runtime/master_session.cc:1004] Start master session 01acd7ab2d13488f with config: gpu_options { allow_growth: true } allow_soft_placement: true
2017-12-13 16:34:48.414965: I tensorflow/core/distributed_runtime/master_session.cc:1004] Start master session ee5f75d36f8fa12e with config: gpu_options { allow_growth: true } allow_soft_placement: true
INFO:tensorflow:global_step/sec: 25.0699
INFO:tensorflow:global_step/sec: 26.0601
INFO:tensorflow:global_step/sec: 24.5106
```